include/mumble/Base64.hpp

```diff
@@ -3,30 +3,134 @@
 // that can be found in the LICENSE file at the root of the
 // Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
+// Inspired by https://renenyffenegger.ch/notes/development/Base64/Encoding-and-decoding-base-64-with-cpp.
+
 #ifndef MUMBLE_BASE64_HPP
 #define MUMBLE_BASE64_HPP
 
-#include "Macros.hpp"
-#include "NonCopyable.hpp"
 #include "Types.hpp"
 
-#include <memory>
+#include <array>
+#include <cmath>
+#include <limits>
+
+#include <gsl/span>
 
 namespace mumble {
-class MUMBLE_EXPORT Base64 : NonCopyable {
+class Base64 {
```

Comment:
```suggestion
namespace base64 {
```
We should avoid using classes as namespaces.

Reply:
We have private members in `Base64`.

Reply:
But we don't need to. If you want "private" definitions, use a nested `details` namespace. That's what is usually done.

Reply:
But why would we want to do that? I mean, what is the actual advantage?

Reply:
Not using classes as namespaces. Classes are meant for defining instantiable objects, but here we don't want anything to be instantiable.

Reply:
It brings a few disadvantages though, such as members having to be declared beforehand and the need to specify that we want to use the `details` namespace in the function(s).

Reply:
I would consider that to be less of a disadvantage than having an instantiable class flying around, that actually isn't meant to be instantiated :thinking:

Reply:
In my opinion it's actually a greater disadvantage, especially considering the compiler should optimize out the allocation of a class that's effectively 0 bytes.

Reply:
This is about semantics, not performance. It literally makes zero sense to define a class that is not meant to be instantiated. What we are doing is 99% what namespaces are for.
Having private definitions isn't possible with namespaces, but what we are currently using it for is also unnecessary. I'm quite sure we can do without casting the span to an int-span, but instead keep using the byte-span.
And what would it hurt to have the pad character and the char to pos functions be public? And if you for some reason really think they should not be part of the public API, put them into a `details` nested namespace as everyone else seems to do as well.
If you are bothered by the fact that this requires defining the private stuff first, then simply outsource them into a separate header file in a `details` subdirectory and include that as the first thing. This will also completely remove those definitions from the eyes of everyone looking at the public API - which the class-approach can't do.

Reply:
> I'm quite sure we can do without casting the span to an int-span, but instead keep using the byte-span.

Already done.

> And what would it hurt to have the pad character and the char to pos functions be public? 

We can make them public now that the data types are consistent, no problem.

> If you are bothered by the fact that this requires defining the private stuff first, then simply outsource them into a separate header file in a `details` subdirectory and include that as the first thing. This will also completely remove those definitions from the eyes of everyone looking at the public API - which the class-approach can't do.

And what if we want to define public stuff in a specific order?

Using a namespace would force us to move `PAD_CHAR`, `charToByte()` and `byteToChar()` at the very top unless we make use of forward declarations.

Also, we may eventually find a case with two (or more) functions depending on each other.

Reply:
> And what if we want to define public stuff in a specific order?

Why would you want to do that? And if you really want to, you can always declare functions first - in whatever order you want - and then  define them later (or in a separate file).

 > Also, we may eventually find a case with two (or more) functions depending on each other.

No problem: first define one and then the other. And if A depends on B and B depends on A, you've most likely introduced shady, if not buggy code anyway.